lib/plugins/plugin.rb

```diff
@@ -0,0 +1,201 @@
+    when 'api'
+      load_api
+    else
+      @logger.error("Plugin: #{@name}: unknown plugin type configured #{@config['plugin']['type']}")
+      raise "Plugin: #{@name}: only 'script', 'container', and 'api' are known"
+    end
+    help_load
+  end
+
+  # Build out help commands for users to query in chat
+  def help_load
+    @help_list = ''
+
+    if @container_hash['Labels']
+      @help_hash = @container_hash['Labels']
+    elsif @api_info.key? :Help
+      @help_hash = @api_info['help']
+    elsif @config['plugin']['help']
+      @help_hash = @config['plugin']['help']
+    end
+
+    @logger.info("#{@name}: #{@help_hash ? 'Help list being built' : 'No help or labels found'}")
+
+    @help_hash&.each do |label, desc|
+      @help_list += '    ' + label + ' : ' + desc + "\n"
+    end
+  end
+
+  def help
+    @help_list
+  end
+
+  # Execute the command sent from chat
+  #
+  # @param string data_from_chat
+  # @return string representing response to be displayed
+  def exec(chat_data = nil)
+    # based on some meta information like the type then execute the proper way
+    # Strip an incorrection coded qoutes to UTF8 ones
+    data_from_chat = sterilize(chat_data)
+    # Split chat data after @bot plugin into args to pass into plugin.
+    # Split args based on qoutes, args are based on spaces unless in qoutes
+    chat_text_array = data_from_chat.text.split(/\s(?=(?:[^"]|"[^"]*")*$)/)
+    exec_data = data_type(data_from_chat, chat_text_array)
+    case @config['plugin']['type']
+    when 'script', 'container'
+      case @config['plugin']['listen_type']
+      when 'passive'
+        exec_passive(exec_data)
+      when 'active'
+        @container.exec([exec_data])
+      end
+    when 'api'
+      exec_api(data_from_chat, chat_text_array)
+    end
+  end
+
+  def exec_passive(exec_data)
+    @container_hash[:Cmd] = exec_data
+    @container = Docker::Container.create(@container_hash)
+    @container.tap(&:start).attach(tty: true)
+    response = @container.logs(stdout: true)
+    @container.delete(force: true)
+    response
+  end
+
+  def exec_api(data_from_chat, chat_text_array)
+    payload = {
+      chat:
+        {
+          user: data_from_chat['user'],
+          channel: data_from_chat['channel'],
+          type: data_from_chat['type'],
+          timestamp: data_from_chat['ts']
+        },
+      command: {
+        # text without username or plugin name
+        data: chat_text_array.drop(2)
+      }
+    }
+    HTTParty.post(@config['plugin']['api']['url'], body: payload, headers: @headers)
+    # else ?
+    # Error log and chat?
+    # Since it will only make it to this level if the bot was invoked
+    # then may it is appropriate to state that the bot does not understand?.
+  end
+
+  # Clears out existing container with the name planned to use
+  # Avoids this error:
+  # Uncaught exception: Conflict. The name "/hello_world" is already in use by container 23ee03db81c93cb7dd9eba206c3a7e.
```

Comment:
I don't think it does, it's does need to be added to test. But I was waiting to use codeclimate after the merge to create the issues (I like one button click solutions)